src/generic/PermGroups.jl

```diff
@@ -0,0 +1,358 @@
+
+doc"""
+    inv(a::perm)
+> Return the inverse of the given permutation, i.e. the permuation $a^{-1}$
+> such that $a\circ a^{-1} = a^{-1}\circ a$ is the identity permutation.
+"""
+function inv(a::perm)
+   d = similar(a.d)
+   for i in 1:length(a.d)
+      d[a[i]] = i
+   end
+   G = parent(a)
+   return G(d)
+end
+
+# TODO: can we do that in place??
+function inv!(a::perm)
+   d = similar(a.d)
+   for i in 1:length(a.d)
+      d[a[i]] = i
+   end
+   a.d = d
+end
+
+###############################################################################
+#
+#   Iterating over all permutations
+#
+###############################################################################
+
+doc"""
+   AllPerms(n::Int)
+> Returns an iterator over arrays representing all permutations of `1:n`.
+> Similar to `Combinatorics.permutations(1:n)`
+"""
+immutable AllPerms
+   n::Int
+   all::Int
+   AllPerms(n::Int) = new(n, factorial(n))
+end
+
+Base.start(A::AllPerms) = (collect(1:A.n), 1, 1, ones(Int, A.n))
+Base.next(A::AllPerms, state) = all_perms(state...)
+Base.done(A::AllPerms, state) = state[2] > A.all
+Base.eltype(::AllPerms) = Vector{Int}
+Base.length(A::AllPerms) = factorial(A.n)
+
+function all_perms(elts, counter, i, c)
+   if counter == 1
+      return (copy(elts), (elts, counter+1, i, c))
+   end
+   n = length(elts)
+   while i <= n
+      if c[i] < i
+         if isodd(i)
+            elts[1], elts[i] = elts[i], elts[1]
+         else
+            elts[c[i]], elts[i] = elts[i], elts[c[i]]
+         end
+         c[i] += 1
+         i = 1
+         return (copy(elts), (elts, counter+1, i, c))
+      else
+         c[i] = 1
+         i += 1
+      end
+   end
+end
+
+doc"""
+    elements(G::PermGroup)
+> Returns an iterator over all elements in the group $G$. You may use
+> `collect(elements(G))` to get an array of all elements.
+"""
+elements(G::PermGroup) = (G(p) for p in AllPerms(G.n))
+
+###############################################################################
+#
+#   Misc
+#
+###############################################################################
+
+doc"""
+    order(G::PermGroup)
+> Returns the order of the full permutation group.
+"""
+order(G::PermGroup) = factorial(G.n)
+
+doc"""
+    cycles(a::perm)
+> Decomposes permutation into disjoint cycles.
+"""
+function cycles(a::perm)
+   to_visit = trues(a.d)
+   cycles = Vector{Vector{Int}}()
+   k = 1
+   while any(to_visit)
+      cycle = Vector{Int}()
+      k = findnext(to_visit, k)
+      to_visit[k] = false
+      push!(cycle, k)
+      next = a[k]
+      while next != k
+         push!(cycle, next)
+         to_visit[next] = false
+         next = a[next]
+      end
+      push!(cycles, cycle)
+   end
+   return cycles
+end
+
+doc"""
+    matrix_repr(a::perm)
+> Return the permutation matrix representing `a` via natural embedding of the
+> permutation group into general linear group over ZZ
+"""
+function matrix_repr(a::perm)
```

Comment:
I see the rational behind this function, but in Nemo matrices are not the same as julia matrices. This will yield `Array{Int, 2}`, but in Nemo matrix functions like `Matrix(ZZ, 1, 2, [1, 2])` yield different types. What do you think @wbhart